panoramapublic/src/org/labkey/panoramapublic/model/speclib/SpectralLibrary.java

```diff
@@ -0,0 +1,248 @@
+                {
+                    _libFilePath = path;
+                    try { _fileSize = Files.size(_libFilePath); } catch (IOException ignored) {}
+                }
+            }
+        }
+    }
+
+    public @NotNull DOM.Renderable getRunLibraryLink(@NotNull User user, @NotNull Map<String, String> viewSpecLibParams)
+    {
+        return SPAN(getRunLink(user),
+                HtmlString.NBSP,
+                SPAN(at(style, "margin-left:5px;"), getViewLibInfoLink(viewSpecLibParams)));
+    }
+
+    public @NotNull DOM.Renderable getRunLink(@NotNull User user)
+    {
+        initRun(user);
+        if (_run == null)
+        {
+            return SPAN(at(style, "color:red;"), "Run not found");
+        }
+        var runUrl = PageFlowUtil.urlProvider(TargetedMSUrls.class).getShowRunUrl(_run.getContainer(), _run.getId());
+        return new Link.LinkBuilder(_run.getFileName()).href(runUrl).clearClasses().build();
+    }
+
+    public @NotNull DOM.Renderable getViewLibInfoAndDownloadLink(@NotNull User user, @NotNull Map<String, String> viewSpecLibParams)
+    {
+        return SPAN(getViewLibInfoLink(viewSpecLibParams), getDownloadLink(user));
+    }
+
+    @NotNull
+    private DOM.Renderable getViewLibInfoLink(@NotNull Map<String, String> viewSpecLibParams)
+    {
+        var viewSpecLibAction = new ActionURL(PanoramaPublicController.ViewSpecLibAction.class, _run.getContainer());
+        viewSpecLibAction.addParameter("specLibId", getId());
+        viewSpecLibParams.forEach(viewSpecLibAction::replaceParameter);
+        return new Link.LinkBuilder("Library").href(viewSpecLibAction).tooltip("View library details").build();
+    }
+
+    @NotNull
+    public DOM.Renderable getDownloadLink(@NotNull User user)
+    {
+        initRun(user);
+        if (_run != null)
+        {
+            initLibPath(user);
+            if (_libFilePath != null)
+            {
+                var webdavUrl = getWebdavUrl(_run.getContainer(), _libFilePath);
+                if (webdavUrl != null)
+                {
+                    var displaySize = FileUtils.byteCountToDisplaySize(_fileSize);
+                    return SPAN(
+                            new Link.LinkBuilder().href(webdavUrl).iconCls("fa fa-download").build(),
+                            HtmlString.NBSP,
+                            new Link.LinkBuilder(displaySize)
+                                    .href(webdavUrl)
+                                    .tooltip("Download library file included in the Skyline document")
+                                    .clearClasses().build()
+                    );
+                }
+                else
+                {
+                    return missingLibrary("Cannot build WebDAV URL");
+                }
+            }
+            return missingLibrary("Library not included in the Skyline document zip file");
+        }
+        return SPAN(at(style, "color:red;"), "Run not found");
```

Comment:
Do you want to use a CSS class for any of these errors? Maybe the existing labkey-error class?

Reply:
Done.